File: py/permute.py
def per(a,cur):
    def swap(a,i,j):
        t = a[i]
        a[i] = a[j]
        a[j] = t
    if cur==len(a) - 1:
        print(a)
        return
    for i in range(cur,len(a)):
        if i!=cur:
            if a[i]!=a[cur]:
                swap(a,i,cur)
                per(a,cur+1)
                swap(a,i,cur)
        else:
                swap(a,i,cur)
                per(a,cur+1)
                swap(a,i,cur)

File: py/permute.java
import java.util.Arrays;

public class permute {
    public static void main(String[] args) {
        int[] a = {4,7,2,9,1};
        dfs(a,0);

    }
    public static void dfs(int[] a, int cur){
        if (cur==a.length-1){
            System.out.println(Arrays.toString(a));
            return;
        }
        for (int i=cur;i<a.length;i++){
            swap(a,i,cur);
            dfs(a,cur+1);
            swap(a,i,cur);
        }
    }
    public static void swap(int[] a,int i,int j){
        int t= a[i];
        a[i] = a[j];
        a[j] = t;
    }
}
